src/EditModel/SaveButton.component.js

```diff
@@ -6,22 +6,41 @@ import addD2Context from 'd2-ui/lib/component-helpers/addD2Context';
 
 function SaveButton(props, { d2 }) {
     const {
+        label,
         isSaving,
         isValid,
         onClick,
         ...rest
     } = props;
-    const buttonText = isSaving ? d2.i18n.getTranslation('saving') : d2.i18n.getTranslation('save');
+
+    const buttonText = label
+        ? label
+        : isSaving 
+            ? d2.i18n.getTranslation('saving') 
+            : d2.i18n.getTranslation('save');
+
     return (
-        <RaisedButton {...rest} primary onClick={onClick} label={buttonText} disabled={isSaving} />
+        <RaisedButton
+            {...rest}
+            primary
+            onClick={onClick}
+            label={buttonText}
+            disabled={isSaving}
+        />
     );
 }
 
 SaveButton.propTypes = {
+    label: PropTypes.string,
     isSaving: PropTypes.bool,
     isValid: PropTypes.bool,
     onClick: PropTypes.func.isRequired,
 };
-SaveButton.defaultProps = { isSaving: false, isValid: true };
+
+SaveButton.defaultProps = {
+    label: undefined,
+    isSaving: false,
```

Comment:
You don't need to say that `label` is `undefined` or that `isSaving` is `false` by default, they will be false-y if omitted.

Reply:
Actually I wanted to set label to empty string.. not undefined.. good that you brought this one up!
I think it's generally better to define default values so we're type safe. Most of the time it probably won't matter, but there are cases where this might lead to problems (also with primitives, like when `.toString` or `.valueOf` is used somewhere down the component tree)..
And as other devs who are working on a component down the tree probably won't check that, I think it's generally a better way to simply define everything that's optional.
What do you think about that?

Reply:
I just try to avoid the redundancy of `Component.defaultProps` when `Component.propTypes` does a better job.

If `SaveButton.propTypes.label` is set to `string.isRequired` then you don't need to provide a `defaultProps.label = ''` since `<SaveButton label=""/>` will satisfy the prop type check.

An empty string is not really a _good default_ value anyway, so at that point might as well just set it to required and let the consumer pass in an empty string if it absolutely must do something so silly.

Setting false-y booleans in `defaultProps` adds a lot of noise to debugging tools since all of the `defaultProps` will be added to the vDOM, e.g. for a button:

```jsx
// with defaultProps set
<Button primary={true} secondary={false} small={true} large={false} destructive={false} dense={false} />

// without defaultProps set
<Button primary small />
```

This adds more noise to the vDOM (and Storybook for component libs) which makes it harder to get an overview of what is _actually_ passed in as a prop and what is _actually_ a default value, and I tend toward the more readable option.

Sometimes `defaultProps` makes sense. In the case above I'd just do:

```jsx
SaveButton.defaultProps = {
    isValid: true,
}

SaveButton.propTypes = {
    label: propTypes.string.isRequired,
    [...]
}
```

Reply:
Just to be clear, I don't think you have to change it (other than `defaultProps.label = ''`). Just noted what I think about it.

Reply:
I newly introduced the `label` prop, which is just used in my case. All other instances of the `SaveButton` don't use it, because it internally determined the label. So setting it to `.isRequired` would break all other usages.

I see the point with the debug tools, I don't get what you mean by this though:
> An empty string is not really a good default value anyway

Could you elaborate what you mean by that? Because to me, the opposite has been true so far (treating empty string as no value always has worked for me)

Reply:
If not using an empty string, some component inside said component might do something like
`const label = props.label + ' *'`, which would result in `"undefined *"`.

Of course in this case this, for both empty string and undefined, this would be a bad case, but there are many cases where `undefined` will case worse effects than an empty string.

This is especially true when a dev works on the component which received the value, which has defaulted in the parent component. The dev probably won't even know which components are used as parents and therefore expects a string.

Reply:
If a component tries to do a blind operation like `const label = props.label + ' *'` the `label` prop should absolutely be set to `.isRequired`. I think that's a better way to ensure correct behaviour and have concise code. Using a default value here just seems like a convenience method over:

```
const label = (props.label || '').concat(' *')
```

Whereas actually requiring `props.label` to be a string is a more "type safe" way to deal with it.

>Could you elaborate what you mean by that? Because to me, the opposite has been true so far (treating empty string as no value always has worked for me)

Yes, it works, no disagreement there.

What I mean is that I don't think an empty string is a _meaningful_ default value. As an example, `label = 'Save'` is a more meaningful default for a string property than `label = ''`.

Reply:
Thanks for the explanation! I agree that empty string should be the last resort when it comes to the default value!